core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1209,8 +1209,11 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, time: Time, met
       // the odds of processing recent broker changes in a single ControllerEvent (KAFKA-5502).
       val curBrokers = zkUtils.getAllBrokersInCluster().toSet
       val curBrokerIds = curBrokers.map(_.id)
+      val brokerIdsOnEventReceived = brokersOnEventReceived.map(_.toInt).toSet
       val liveOrShuttingDownBrokerIds = controllerContext.liveOrShuttingDownBrokerIds
-      val newBrokerIds = curBrokerIds -- liveOrShuttingDownBrokerIds
+      // Also call onBrokerStartup for brokers in (curBrokerIds -- brokerIdsOnEventReceived) to take care of the
+      // scenario that the broker has started after the event is received but before the event is processed
+      val newBrokerIds = (curBrokerIds -- liveOrShuttingDownBrokerIds) ++ (curBrokerIds -- brokerIdsOnEventReceived)
```

Comment:
I wonder if the following expression suffices:
```
(curBrokerIds -- liveOrShuttingDownBrokerIds) -- brokerIdsOnEventReceived
```

Reply:
@tedyu Thanks for the comment. I will improve it as you suggested.

Reply:
@tedyu I just double checked it again. Your suggestion would not work. The counter example is curBrokerIds=1,2,3. liveOrShuttingdownBrokerId=1,2,3 brokerIdsOnEventReceived=1,2

Reply:
And this is why I originally wrote it in that way.